core/src/main/java/org/apache/calcite/schema/impl/ListTransientTable.java

```diff
@@ -99,6 +99,9 @@ public ListTransientTable(String name, RelDataType rowType) {
           // TODO cleaner way to handle non-array objects?
           @Override public Object[] current() {
             Object current = list.get(i);
+            if (current == null) {
+              return null;
+            }
```

Comment:
When the `current` can be null ?

Reply:
ListTransientTable is currently used in combination with repeatUnion to support recursive queries. So we could get to this scenario if we have a null involved in the repeatUnion.

Reply:
Hmm, is it possible to write a test case there, for your case `null involved repeatUnion`

Reply:
Thanks for the feedback @danny0405 . I have provided a unit test that reproduces the NPE.